text/0000-build-observability.md

```diff
@@ -0,0 +1,358 @@
+- Author(s): @joshwlewis
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes leveraging [OpenTelemetry](https://opentelemetry.io/) to
+grant platform operators and buildpack operators more insight into buildpack
+performance and behavior. This RFC describes new opt-in functionality
+for both pack and the buildpack spec such that OpenTelemetry data may be
+exported to the build file system.
+
+# Definitions
+[definitions]: #definitions
+
+- [OpenTelemetry](https://opentelemetry.io/): A collection of APIs, SDKs, and tools that can be used it to instrument, generate, collect, and export telemetry data.
+- [Traces](https://opentelemetry.io/docs/concepts/signals/traces/): Telemetry
+  category that describes the path of software execution.
+
+
+# Motivation
+[motivation]: #motivation
+
+Buildpack authors and platform operators desire insight into usage, error
+scenarios, and performance of builds and buildpacks on their platform. The
+following questions are all important for these folks, but difficult to answer:
+
+- "Which buildpacks commonly fail to compile?"
+- "How often does a particular error scenario occur?"
+- "How long does each buildpack compile phase take?"
+- "How often is a certain buildpack used?"
+- "Which versions of Go are being installed?"
+- "How long does it take to download node_modules?"
+
+Instrumenting lifecycle and buildpacks with opt-in OpenTelemetry tracing will
+allow  platform operators to better understand performance and behavior of their
+builds and buildpacks and as a result, provide better service and build
+experiences.
+
+To protect privacy and prevent unnecessary collection of data, this
+functionality should be optional and anonymous.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC aims to provide a solution for two types of OpenTelemetry traces:
+
+1) Lifecycle tracing: Buildpack-agnostic trace data like which buildpacks were
+available, which buildpacks were detected, how long the detect, build, or
+export phase took, and so on. This telemetry data may be exported by lifecycle.
+2) Buildpack tracing: Telemetry data specific to a buildpack like how long it
+took to download a language binary, which language version was selected, and so
+on. This telemetry data may be exported by buildpacks.
+
+Though the sources and contents of the telemetry data differ, both types may
+be emitted to the build file system in OpenTelemetry's [File Exporter
+Format](https://opentelemetry.io/docs/specs/otel/protocol/file-exporter/).
+
+In this solution, each lifecycle phase would write a `.jsonl` file with
+tracing data for that phase. For example, `lifecycle detector --telemetry`
+would write to `/cnb/telemetry/lifecycle-detect.jsonl`. Additionally each
```

Comment:
Right now we use the layers directory as the de-facto place for "output" (layers/report.toml, layers/sbom, etc.). Would it be easier to find a new place under layers?

That said adding yet another output directory makes me want to look at something like https://github.com/buildpacks/rfcs/pull/145 again. Not that we need to block this RFC on that, but maybe it's worth revisiting in tandem.

Reply:
Yeah, I thought about `/layers`. I didn't choose it because I don't think it's available during some of the earlier lifecycle phases (like `detect`)? Maybe I should confirm that - I know at the very least, the `/layers` subdirectories don't exist until later phases.

`/cnb/telemetry` is definitely a strawman - I'm not committed to it at all.

Reply:
I've updated the file locations to `/layers`.